File: frontend/src/types/foreman.ts
// Foreman type definitions — aligned with backend models.

// ============ Agent directory ============

export type ForemanAgentPresence = "online" | "busy" | "offline";

export interface ForemanAgentDirectoryItem {
  /** Format: "business_unit_id/agent_name" */
  id: string;
  businessUnitId: string;
  businessUnitName: string;
  name: string;
  displayName: string;
  description: string;
  /** Derived on frontend side; backend doesn't track presence yet. */
  presence: ForemanAgentPresence;
  accentColor: string;
}

// ============ Conversation member ============

export interface ForemanConversationMember {
  memberId: string;
  businessUnitId: string;
  agentName: string;
  displayName: string;
  joinedAt: string;
}

// ============ Conversation types ============

export type ForemanConversationType = "direct" | "group";

export interface ForemanConversationSummary {
  conversationId: string;
  conversationType: ForemanConversationType;
  title: string;
  lastMessagePreview: string;
  memberCount: number;
  updatedAt: string;
  createdAt: string;
}

export interface ForemanConversationDetail {
  conversationId: string;
  conversationType: ForemanConversationType;
  title: string;
  lastMessagePreview: string;
  coordinatorEnabled: boolean;
  members: ForemanConversationMember[];
  updatedAt: string;
  createdAt: string;
}

// ============ Timeline message ============

export type ForemanMessageRole = "user" | "agent" | "system";

/** Execution step status for task list display. */
export type ForemanTaskStatus = "pending" | "running" | "completed" | "failed" | "cancelled";

/** One step in the agent execution plan. */
export interface ForemanTaskStep {
  id: string;
  title: string;
  description?: string;
  status: ForemanTaskStatus;
  /** Elapsed time in seconds for display (e.g. "00:02"). */
  elapsedSec?: number;
  /** Thought/reasoning content for this step (expandable). */
  thought?: string;
}

/** Tool call entry tracked during agent execution. */
export interface ForemanToolCallEntry {
  toolCallId?: string;
  toolName: string;
  toolArgs?: Record<string, any>;
  toolResult?: string;
  status: "running" | "completed" | "failed";
  reason?: string;
  expectedOutcome?: string;
  reflection?: string;
}

export interface ForemanMessage {
  /** Primary identifier for the message. */
  id: string;
  /** Same as id; used by backend contracts. */
  messageId: string;
  conversationId: string;
  roundId?: string;
  role: ForemanMessageRole;
  senderId: string;
  senderName: string;
  /** Accent color for the agent avatar. */
  senderColor?: string;
  content: string;
  mentionedAgents?: string[];
  createdAt: string;

  // ---- Execution process fields (agent messages only) ----

  /** Whether the agent is currently executing (show loading indicators). */
  isExecuting?: boolean;
  /** Whether the thinking process panel is expanded. */
  isThinkingExpanded?: boolean;
  /** Current thought/reasoning text being streamed. */
  thoughtText?: string;
  /** Current execution phase label. */
  currentPhase?: string;
  /** Execution step list. */
  tasks?: ForemanTaskStep[];
  /** Tool call entries. */
  toolCalls?: ForemanToolCallEntry[];
  /** Done summary text. */
  doneSummary?: string;
  /** Suggested next steps after completion. */
  doneNextSteps?: string[];
  /** Error info. */
  errorText?: string;
}

// ============ Sidebar item (unified for left panel) ============

export type ForemanSidebarItemType = "agent" | "group";

export interface ForemanSidebarItem {
  id: string;
  type: ForemanSidebarItemType;
  title: string;
  summary: string;
  updatedAt: string;
  agentIds: string[];
  businessUnitName?: string;
  accentColor?: string;
  presence?: ForemanAgentPresence;
}

// ============ SSE event types ============

export type ForemanSSEEventKind =
  | "round_started"
  | "member_started"
  | "member_done"
  | "round_completed"
  | "coordinator_decision";

// ============ Local state (for useForemanChat) ============

/**
 * Runtime conversation state held in memory by useForemanChat.
 * Combines backend detail with locally-accumulated messages.
 */
export interface ForemanConversation {
  id: string;
  type: ForemanConversationType;
  title: string;
  summary: string;
  agentIds: string[];
  messages: ForemanMessage[];
  updatedAt: string;
}

export interface ForemanPersistedState {
  conversations: ForemanConversation[];
  activeConversationId: string | null;
}

File: backend/app/models/foreman.py
"""Foreman conversation models.

Defines the data contracts for Foreman multi-agent chat: conversations,
members, timeline messages, and request/response schemas used by the
``/api/foreman`` endpoints.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import List, Literal, Optional

from pydantic import BaseModel, Field


# ============================================================
# Agent directory item (returned by GET /api/foreman/agents)
# ============================================================

class ForemanAgentInfo(BaseModel):
    """Lightweight agent descriptor for the Foreman directory listing."""

    id: str = Field(..., description="Unique agent identifier (business_unit_id + agent_name)")
    business_unit_id: str
    business_unit_name: str
    name: str = Field(..., description="Agent directory name")
    display_name: str
    description: Optional[str] = None


# ============================================================
# Conversation member
# ============================================================

class ConversationMember(BaseModel):
    """One agent member inside a Foreman conversation."""

    member_id: str = Field(default_factory=lambda: str(uuid.uuid4())[:12])
    business_unit_id: str
    agent_name: str
    display_name: str
    joined_at: str = Field(default_factory=lambda: datetime.now().isoformat())


# ============================================================
# Conversation models
# ============================================================

class ConversationSummary(BaseModel):
    """Lightweight conversation descriptor for the session list."""

    conversation_id: str
    conversation_type: Literal["direct", "group"] = "direct"
    title: str
    last_message_preview: str = ""
    member_count: int = 1
    updated_at: str = ""
    created_at: str = ""


class ConversationDetail(BaseModel):
    """Full conversation payload including members."""

    conversation_id: str
    conversation_type: Literal["direct", "group"] = "direct"
    title: str
    last_message_preview: str = ""
    coordinator_enabled: bool = False
    members: List[ConversationMember] = Field(default_factory=list)
    updated_at: str = ""
    created_at: str = ""


# ============================================================
# Timeline message
# ============================================================

class TimelineMessage(BaseModel):
    """One entry in the unified conversation timeline."""

    message_id: str = Field(default_factory=lambda: str(uuid.uuid4())[:12])
    conversation_id: str
    round_id: Optional[str] = None
    role: Literal["user", "agent", "system"] = "user"
    sender_id: str = ""
    sender_name: str = ""
    content: str = ""
    mentioned_agents: Optional[List[str]] = None
    created_at: str = Field(default_factory=lambda: datetime.now().isoformat())


# ============================================================
# Request schemas
# ============================================================

class CreateConversationRequest(BaseModel):
    """Create a new single-chat or group-chat conversation."""

    agent_ids: List[str] = Field(
        ...,
        min_length=1,
        description="List of agent identifiers (business_unit_id/agent_name format)",
    )
    title: Optional[str] = None


class AddMembersRequest(BaseModel):
    """Add one or more agents to an existing conversation."""

    agent_ids: List[str] = Field(
        ...,
        min_length=1,
        description="Agent identifiers to add",
    )


class SendMessageRequest(BaseModel):
    """User message sent into a conversation stream."""

    content: str = Field(..., min_length=1)
    mentions: Optional[List[str]] = None


# ============================================================
# Response helpers
# ============================================================

class ConversationListResponse(BaseModel):
    """Wrapper for the conversation list endpoint."""

    conversations: List[ConversationSummary]


class TimelineResponse(BaseModel):
    """Paginated timeline response."""

    conversation_id: str
    messages: List[TimelineMessage]
    total: int = 0
